accepted/0000-caasp-integration.md

```diff
@@ -0,0 +1,395 @@
+Hardware details of the node will be shown.
+
+From a Uyuni/SUSE Manager point of view, it would make sense to create a system group to identify the nodes part of the cluster.
+
+#### Show cluster status
+
+Uyuni/SUSE Manager shows the status of the cluster by invoking `skuba cluster status` on the Management Node (via Salt).
+
+E.g. The output is presented in the "CaaS Platform > cluster name" section:
+
+```
+NAME         OS-IMAGE                              KERNEL-VERSION        CONTAINER-RUNTIME   HAS-UPDATES   HAS-DISRUPTIVE-UPDATES
+cp0       SUSE Linux Enterprise Server 15 SP1   4.12.14-110-default   cri-o://1.13.3      <none> <none>
+worker0   SUSE Linux Enterprise Server 15 SP1   4.12.14-110-default   cri-o://1.13.3      <none> <none>
+```
+
+#### Removing nodes from the cluster
+
+A node can also be removed from the cluster:
+
+- temporarily (drained and cordoned off): [using `kubectl`](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#use-kubectl-drain-to-remove-a-node-from-service) from the Management Node via Salt.
+- permanently: by invoking `skuba node remove <nodename>` via Salt. NOTE: this removal is permanent and the cluster cannot join any cluster in the future. A complete reinstallation is required and Uyuni/SUSE Manager must offer the user to re-provision the node from scratch, e.g. via AutoYaST.
+
+Example state:
+
+```
+remove_node:
+      cmd.run:
+        - name: skuba node remove <nodename>
+```
+
+NOTE: Other `skuba` options can be implemented in the future (e.g. resetting and reconfiguring a node).
+
+#### Cluster upgrade
+
+All the cluster updates must be handled by `skuba`.
+
+Uyuni/SUSE Manager can check and notify the user if an update for the cluster is present by triggering `skuba cluster upgrade plan` on the Management Node (using Salt) and check if a new output is available in a cron fashion (e.g. Taskomatic job).
+
+Uyuni/SUSE Manager will then interface with the Management node and upgrade every control plane in the cluster first:
+
+`skuba node upgrade apply --target <control-plane-node-ip> --user <user> --sudo`
+
+and then every worker:
+
+`skuba node upgrade apply --target <worker-node-ip> --user <user> --sudo`
+
+This can be done by supplying the command to the Management Node via Salt.
+
+NOTE: CaaS Platform stack is also deployed via containers. We will need to update this section when the CaaS Platform team concludes the "add-ons update story".
+
+#### CVE Auditing
+
+By leveraging the existing feature of CVE Auditing, Uyuni/SUSE Manager can detect and notify the user whether the cluster is running a vulnerable image. When Uyuni/SUSE Manager rebuilds the image and patches it, Uyuni/SUSE Manager can also guide the user in modifying the Kubernetes deployment to pull the updated image (in case of `imagePullPolicy:` is not `Always`) or do a `kubectl rolling-update`.
```

Comment:
This is open to discussion: we can offer two strategies (just warn, refuse to start) and the user can select one of those. 
What features do you need for CVE in SOC other than auditing?

Reply:
Audit is nice.
We can handle the annotation of pods/rejecting the pods ourselves if necessary. We should discuss this in our community of architecture with our PM.
For the automatic image building and update of manifests, I would prefer to opt out until we know more about how you will do this.

Reply:
> This seems very disruptive by default:

But that's exactly where larger organizations see the added benefit of tools like SUSE Manager. Because they _want_ to add another layer of control and not just use vendor-provided packages or images untested.

> Should we really assume that the customer will have to "patch" the images, or should we make sure, in the SUSE org, to have a proper process to deliver updated/secure images in a timely fashion?

Of course this doesn't mean that we should _force_ anyone into not using the latest images from SUSE.

Also, in most cases the customer won't actually re-build images, but just hold them back until they've been tested.

The re-building option for images that belong to infrastructure (as opposed to the customer's specific workloads) is really only required when customer-specific adjustments are needed. And most of those would require a special supportability agreement.

> I think this needs to be decided per product. I expect OpenStack delivery of releases could take a long while, but letting the customer update the images without coordinated approach sounds like the source of headaches and downtime.

I'm not sure whether "per product" makes any sense here. Customers are going to install full or partial stacks from SUSE. If we even have different policies between parts of those supposedly integrated stacks, we won't make users happy.

That said, of course customers shouldn't be doing anything in an _uncoordinated_ way. Usually, the use case here is that a certain stack that has been released together is held back as a whole until it's been tested in the customer's own QA environment.

> If we still assume patching of the image, how can we ensure this will be in sync with the manifests of the deployment? How does that map to gitops?

That's a very good point. We _are_ looking into making SUSE Manager more GitOps-friendly across the board. How exactly we're going to get the build and configuration stories fully integrated, so that both can be driven from one set of Git-managed templates, is still TBD.

> ... If it's manual (let's say I built my own image "a", tag "b.c"), will I have to manually update "a" in all my manifests to point to "b.c"? Isn't that the value of SUSE to provide this kind of "grunt work"?

Right. That's exactly where we need to do more research.

> ... If it's automatic, do we have to implement mutating webhooks?
> All of this seems to be slippery slope of extra work.

It isn't easy, but we've had very strong feedback from the field for Kubernetes (and also OpenStack) that not everyone is ready for automation with no control (i.e. latest vendor patches are just installed automatically).

> It looks nice on paper to be able to change the containers, but I doubt we should provide, at least for the cloud team, more than just auditing.

I'd bet that some of the OpenStack customers will be among the first who want more control. :-)

Reply:
I added a small section describing the two strategies that the user can select.

Reply:
> But that's exactly where larger organizations see the added benefit of tools like SUSE Manager. Because they _want_ to add another layer of control and not just use vendor-provided packages or images untested.

Wow that escalated quickly. Do you really mean we don't test images or packages? I sure hope you meant _in the customer environment_. :)

I agree on the fact that customers will love to determine when to do things, and how. But it doesn't mean the atomicity is the image (not that they don't care about the image, but what they care is having something that solve their business need first... in other words, the app!)

When you are deploying openstack you have probably hundreds of images overrides possible. I seriously doubt customers want to deal with this. Some might need to. But most likely all want to be able to determine what's in their image, and when/how they deploy it _as part of a whole_.

IMHO, SUMA benefit for most customers in this spec is that it could help deploying clusters for their applications. Not that it provides little tweaks for images that could break their environment because they are special.

> The re-building option for images that belong to infrastructure (as opposed to the customer's specific workloads) is really only required when customer-specific adjustments are needed. And most of those would require a special supportability agreement.

Thanks for that clarification. I am not sure if that's clear on the OpenStack side. So when I read this, I assume that the feature can be disabled on a case by case basis, based on presence of said agreement.

> I'm not sure whether "per product" makes any sense here. Customers are going to install full or partial stacks from SUSE. If we even have different policies between parts of those supposedly integrated stacks, we won't make users happy.

I seriously doubt that. As a customer, why would I want to follow SUMA's way if it's different than the docs of the app I want to deploy? I would rather follow the docs of the product I want to install on my infrastructure. And that might not be a SUSE product. Do we really want to be closed instead of open? Or did I get your message wrong?

> That said, of course customers shouldn't be doing anything in an _uncoordinated_ way. Usually, the use case here is that a certain stack that has been released together is held back as a whole until it's been tested in the customer's own QA environment.

That proves my opinion that, at best, this should be dealing with clusters, and responsibility of the app is up to the customer. That's the atomicity the customer is looking for. The customer can drill down in the image building (optionally), but I then expect that _for simplicity_ there is no override of the image tag.

> It isn't easy, but we've had very strong feedback from the field for Kubernetes (and also OpenStack) that not everyone is ready for automation with no control (i.e. latest vendor patches are just installed automatically).

I agree, delivery of the "apps" is to be delivered with some user control.

I have the impression we are agreeing in some concepts, but using very different words, which might be the cause of confusions. I would love to see a limit to this spec's scope, and I believe the limit is the cluster. The app is then user's decision, and what we could do is provide "examples" of apps as part of our products. That would allow full customization and control. But I think we are in agreement there, am I right?

Reply:
> > But that's exactly where larger organizations see the added benefit of tools like SUSE Manager. Because they _want_ to add another layer of control and not just use vendor-provided packages or images untested.
> 
> Wow that escalated quickly. Do you really mean we don't test images or packages? I sure hope you meant _in the customer environment_. :)

I'm sure Joe was meaning the customer's environment; that is what I'm getting from the field as well.

> I have the impression we are agreeing in some concepts, but using very different words, which might be the cause of confusions. I would love to see a limit to this spec's scope, and I believe the limit is the cluster. The app is then user's decision, and what we could do is provide "examples" of apps as part of our products. That would allow full customization and control. But I think we are in agreement there, am I right?

Yes, I added a sentence that SUMA's scope is only pertaining to the cluster, not to what's on top of it.